### PR TITLE
Simple fix for memory leak on GPU0

### LINKED ISSUE #958: Process runs on more GPUs than specified
I have a single 8-GPU machine with a faulty GPU0. 
I'm running imagenet_example.py on 7 GPUs on this machine by specifying `gpus=[1,2,3,4,5,6,7]` in the Trainer i.e. I do not want to use GPU0

However, when i run `nvidia-smi`, I see the Trainer's pid shows on all 8 GPUs, just with lower memory on GPU0 (see output below). I also find it to be slower than non-PL code by about 4x. I don't see this behavior if I manually set `CUDA_VISIBLE_DEVICES=1,2,3,4,5,6,7` followed by `gpus=7` in Trainer. Similarly, it works fine when using a single GPU with, say, `gpus=[1]`.
I'm not sure if it's relevant but I also see `gpu=0` in the tqdm progress bar

#### nvidia-smi with Trainer(gpus=[1,2,3,4,5,6,7]) and CUDA_VISIBLE_DEVICES unset
```
+-----------------------------------------------------------------------------+
| Processes:                                                       GPU Memory |
|  GPU       PID   Type   Process name                             Usage      |
|=============================================================================|
|    0     40155      C   python                                       719MiB |
|    1     40155      C   python                                      6003MiB |
|    2     40155      C   python                                      6019MiB |
|    3     40155      C   python                                      6019MiB |
|    4     40155      C   python                                      6019MiB |
|    5     40155      C   python                                      6019MiB |
|    6     40155      C   python                                      6019MiB |
|    7     40155      C   python                                      6019MiB |
+-----------------------------------------------------------------------------+
```

#### nvidia-smi with Trainer(gpus=7) and CUDA_VISIBLE_DEVICES=1,2,3,4,5,6,7

```
+-----------------------------------------------------------------------------+
| Processes:                                                       GPU Memory |
|  GPU       PID   Type   Process name                             Usage      |
|=============================================================================|
|    1     34452      C   python                                      6003MiB |
|    2     34452      C   python                                      6019MiB |
|    3     34452      C   python                                      6019MiB |
|    4     34452      C   python                                      6019MiB |
|    5     34452      C   python                                      6019MiB |
|    6     34452      C   python                                      6019MiB |
|    7     34452      C   python                                      6019MiB |
+-----------------------------------------------------------------------------+
```

### Expected behavior

The process should run on the specified GPUs without manually setting CUDA_VISIBLE_DEVICES

### Environment
```
PyTorch version: 1.4.0
Is debug build: No
CUDA used to build PyTorch: 10.1

OS: Ubuntu 18.04.3 LTS
GCC version: (Ubuntu 7.4.0-1ubuntu1~18.04.1) 7.4.0
CMake version: version 3.10.2

Python version: 3.8
Is CUDA available: Yes
CUDA runtime version: Could not collect
GPU models and configuration:
GPU 0: GeForce RTX 2080 Ti
GPU 1: GeForce RTX 2080 Ti
GPU 2: GeForce RTX 2080 Ti
GPU 3: GeForce RTX 2080 Ti
GPU 4: GeForce RTX 2080 Ti
GPU 5: GeForce RTX 2080 Ti
GPU 6: GeForce RTX 2080 Ti
GPU 7: GeForce RTX 2080 Ti

Nvidia driver version: 418.87.00
cuDNN version: Could not collect

Versions of relevant libraries:
[pip] numpy==1.18.1
[pip] pytorch-lightning==0.6.0
[pip] torch==1.4.0
[pip] torch-lr-finder==0.1.2
[pip] torchvision==0.5.0
[conda] blas                      1.0                         mkl
[conda] mkl                       2020.0                      166
[conda] mkl-service               2.3.0            py38he904b0f_0
[conda] mkl_fft                   1.0.15           py38ha843d7b_0
[conda] mkl_random                1.1.0            py38h962f231_0
[conda] pytorch                   1.4.0           py3.8_cuda10.1.243_cudnn7.6.3_0    pytorch
[conda] pytorch-lightning         0.6.0                    pypi_0    pypi
[conda] torch-lr-finder           0.1.2                    pypi_0    pypi
[conda] torchvision               0.5.0                py38_cu101    pytorch
```

### PR DESCRIPTION
**Pull request by shubhamagarwal92**
# Before submitting

- [X] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [X] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [ ] Did you make sure to update the docs?   
- [ ] Did you write any new necessary tests?  
- [ ] If you made a notable change (that affects users), did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CHANGELOG.md)?

## What does this PR do?
Fixes #958 related to memory leak. Pretty old PyTorch issue related to setting device. 
1. Set the device directly while finding the root gpu in `determine_root_gpu_device` in trainer.distrib_parts.py
2. Remove also the hack from [trainer.evaluation_loop](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/pytorch_lightning/trainer/evaluation_loop.py#L424)

## PR review    
Anyone in the community is free to review the PR once the tests have passed.     
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?
Make sure you had fun coding 🙃 lol. Yeah I did!

### CODE REVIEW

**Comment by shubhamagarwal92** [comment]
@williamFalcon Could you suggest why these tests are failing.

**Comment by awaelchli** [comment]
@shubhamagarwal92 likely because you are calling torch.cuda.* when cuda is not available or test expects it to run on cpu.

**Comment by shubhamagarwal92** [comment]
> @shubhamagarwal92 likely because you are calling torch.cuda.* when cuda is not available or test expects it to run on cpu.

@awaelchli I am assuming that we would `determine_root_gpu_device` only when cuda is available. Anyways, I have added a simple check `root_gpu >= 0` before setting the device in `torch.cuda.set_device()`. Omitting this usually causes this memory leak on GPU0 as suggested by Soumith [here](https://github.com/pytorch/pytorch/issues/9871#issuecomment-408304190). I am not sure if this is the right place to set the device though.

**Comment by Borda** [comment]
hey there, we have added GPU CI test, so could we kindly ask to rebase/merge master which will trigger these tests so we do not need to test it manually... Thx for your understanding :robot:

**Comment by shubhamagarwal92** [comment]
@Borda I have rebased/merged the master!

**Comment by Borda** [comment]
@shubhamagarwal92 could you please rather rebase on master because now it shows all recent master edits and it had to check your changes...
it seems that there some issue while asking for nb GPU on no-GPU machine

**Comment by shubhamagarwal92** [comment]
@Borda I don't know why I am having multiple replicas of the same commit when I am trying to merge this repo's master. For now, I have force-pushed the commit before merge. 

Basically, in this PR, I am trying to set device [here](https://github.com/PyTorchLightning/pytorch-lightning/pull/1094/files#diff-d673a98c14aa5cd59610bcc447178852R648)

File: pytorch_lightning/trainer/distrib_parts.py [hunk 1]
```diff
@@ -640,4 +640,11 @@ def determine_root_gpu_device(gpus):
     # set root gpu
     root_gpu = gpus[0]
 
+    # set cuda device to root gpu
+    # related to https://github.com/PyTorchLightning/pytorch-lightning/issues/958
+    # Refer solution: https://github.com/pytorch/pytorch/issues/9871#issuecomment-408304190
+    # root_device = torch.device("cuda", root_gpu)
```
Comment (jeffling) [inline pytorch_lightning/trainer/distrib_parts.py]: We can remove this now

File: pytorch_lightning/trainer/distrib_parts.py [hunk 2]
```diff
@@ -640,4 +640,11 @@ def determine_root_gpu_device(gpus):
...
+    root_device = (torch.device("cuda", root_gpu) if root_gpu >= 0 else torch.device("cpu"))
```
Comment (jeffling) [inline pytorch_lightning/trainer/distrib_parts.py]: What would root_device be set if the user wants CPU? None? -1? Maybe we should check for that explicitly

Comment (shubhamagarwal92) [reply pytorch_lightning/trainer/distrib_parts.py]: If the user wants CPU, the function `determine_root_gpu_device` should not be called (?)

Comment (awaelchli) [reply pytorch_lightning/trainer/distrib_parts.py]: in this case it is getting called with gpus=None, and returns None (see first lines of determine_root_gpu_device). So your `else torch.device("cpu"))` is never relevant.
I don't see anything wrong with just 
`root_device = torch.device("cuda", root_gpu)`

Comment (awaelchli) [reply pytorch_lightning/trainer/distrib_parts.py]: but I think the device should be set outside this function anyway

Comment (shubhamagarwal92) [reply pytorch_lightning/trainer/distrib_parts.py]: @awaelchli where do you suggest?

**Comment by williamFalcon** [comment]
@Borda is this dead?

**Comment by Borda** [comment]
@shubhamagarwal92 @awaelchli how is it going?

File: pytorch_lightning/trainer/distrib_parts.py (hunk 2 above)
Comment (awaelchli) [reply pytorch_lightning/trainer/distrib_parts.py]: I would search the code base for occurrences of `self.root_gpu` and check whether it is needed to set the device in each case. 
Maybe consider setting directly after [here](https://github.com/PyTorchLightning/pytorch-lightning/blob/a707d4bea1a78a98265fd1ea5b7a7a6cadc37fb9/pytorch_lightning/trainer/trainer.py#L385) but I am not sure if that's the best place.

Comment (awaelchli) [reply pytorch_lightning/trainer/distrib_parts.py]: better to ask the core team on this :)

**Comment by awaelchli** [comment]
If I understand correctly, the fix to the memory leak is simply to add a `torch.cuda.set_device(root_device)` before training. The question is where is the best place to put it. I made a suggestion (see comment above).

**Comment by shubhamagarwal92** [comment]
@awaelchli @Borda @williamFalcon 

I have incorporated the suggestions. See the latest commit https://github.com/PyTorchLightning/pytorch-lightning/pull/1094/commits/6cba621496fd90f2f0f90980e9a425eddebc5c18 

Could you please have a look. Thanks.

**Comment by Borda** [comment]
seems to fail on returned None:
```
        root_device = (torch.device("cuda", self.root_gpu)
>                      if self.root_gpu >= 0 else torch.device("cpu"))
E       TypeError: '>=' not supported between instances of 'NoneType' and 'int'
```

**Comment by williamFalcon** [comment]
@shubhamagarwal92 this is great. mind fixing the issue so we can get into 0.7.2? :)

**Comment by shubhamagarwal92** [comment]
@williamFalcon I updated the if-check in https://github.com/PyTorchLightning/pytorch-lightning/pull/1094/commits/2c7b80265779f8cd702b10ef1386a1c263ad2681

but I do not understand why the automated checks are failing, such as this requires a cuda device:
https://github.com/PyTorchLightning/pytorch-lightning/pull/1094/checks?check_run_id=549408374#step:9:252 

I can only tell that this is probably not the right place to set the device (?) Any suggestions?

File: pytorch_lightning/trainer/evaluation_loop.py [hunk 3]
```diff
@@ -420,10 +420,16 @@ def evaluation_forward(self, model, batch, batch_idx, dataloader_idx, test_mode:
...
         # single GPU data transfer
         if self.single_gpu:
-            # for single GPU put inputs on gpu manually
-            root_gpu = 0
+
             if isinstance(self.data_parallel_device_ids, list):
                 root_gpu = self.data_parallel_device_ids[0]
+                root_device = (torch.device("cuda", root_gpu)
+                               if root_gpu else torch.device("cpu"))
+                torch.cuda.set_device(root_device)
```
Comment (williamFalcon) [inline pytorch_lightning/trainer/evaluation_loop.py:428]: also need to add tpu device...

**Comment by williamFalcon** [comment]
@shubhamagarwal92  that's not where the device should be set. i put it in the correct place in #1349

**Comment by shubhamagarwal92** [comment]
> @shubhamagarwal92 that's not where the device should be set. i put it in the correct place in #1349.
> 
> Let's just merge that one. I think it will still credit you as a co-author

Thanks for doing this. Ahh, now I see where the issue was. Should we close this PR then?

**Pull request closed without merging**